src/qibo/derivative.py

```diff
@@ -0,0 +1,125 @@
+# -*- coding: utf-8 -*-
+def parameter_shift(
+    circuit, observable, parameter_index, gate_eigenv, initial_state=None
+):
+    """In this method the parameter shift rule (PSR) is implemented.
+    Given a circuit U and an observable H, the PSR allows to calculate the derivative
+    of the expected value of H on the final state obtained by application of U with
+    respect to a variational parameter of the circuit.
+
+    Original references:
+        `https://arxiv.org/abs/1811.11184`;
+        `https://arxiv.org/abs/1803.00745`.
+
+    Args:
+        circuit (:class:`qibo.models.circuit.Circuit`): custom quantum circuit.
+        observable (:class: `qibo.hamiltonians.Hamiltonian`): the hamiltonian here is the target observable.
+        parameter_index (int): the index which identify the targe parameter in the circuit.get_parameters list
+        gate_eigenv (float): abs(eigenvalue) of H. In case of Pauli 1/2{sigmas} observable r = 0.5
```

Comment:
I am not sure if this argument is valid here, but I generally don't like duplication in function arguments. Here there is some duplication because one has to pass the observable and the generator eigenvalues seperately, while both can be calculated if only the generator is given to begin with.

If this algorithm only works for observables of the form exp(-iG), maybe it would be cleaner if we take G as an argument and calculate the observable and eigenvalues from it? Note that G can also be a `Hamiltonian` and one can use `ham.exp()` and `ham.eigenvalues()` for these calculations.

That being said, I understand why the observable may be more natural argument as this is what appears in the loss. @DiegoGM91 and other potential users of this, let us know your opinion. In this case we could even think a way to derive the eigenvalues from the observable directly.

What I am thinking is to make `gate_eigenv` optional, something like:
```py
if gate_eigenv is None:
    gate_eigenv = # calculate the eigenvalues automatically if user did not give them
```

Reply:
This specific implementation is valid only for circuits in which the variational parameter appears in a gate whose generator has at most two eigenvalues. With this in mind, perhaps it would converge:

1. add a `raise_error` that rejects ports that do not meet these prerequisites;
2. make a distinction between generic gates and rotations. I say this because a large proportion of the variational quantum algorithms use rotations, and can be useful to have a smart way for involving gates like them (without calculating the eigenvals every time).

What do you think @DiegoGM91 and @stavros11 ?

Reply:
I agree with both points. 

i) For the time being (before we implement generalized parameter shift rules valid for all our gates), we can raise an error. As I commented below, we could remove the `gate_eigenv` argument and use the `parameter_index` to get the type of gate from the queue of gates of the `circuit`. Then, depending on the type of gate, we raise the error.

ii) We already know what the eigenvalues of Pauli matrices are, so I see no point in calculating them every single time.


By the way, point ii) applies not just to single-qubit Paulis, but also to any tensor product of Paulis; this made me think that perhaps we could implement a gate that is the exponential of a tensor product of Paulis times a variational parameter, I think it would be useful. These gates are used, for instance in the Hamiltonian Variational Ansatz. Currently, to implement such a unitary we can use a `hamiltonian` and then the `StateEvolution` model. But what if I want to easily combine such exponential with other built-in gates in `qibo`? Another advantage of having these gates would be that the parameter-shift rule implemented in this PR would apply to them as well.

Reply:
Thank you for the replies and @DiegoGM91 for explaining that the eigenvalues are related to the circuit, not the observable. 

I also agree with raising an error, we should certainly do that for all circuits not supported by the current implementation. If we implement more general rules in the future we will replace the error by delegating to the appropriate rule.

> ii) We already know what the eigenvalues of Pauli matrices are, so I see no point in calculating them every single time.
s well.

What I wrote above is one way (maybe not the best) of implementing this, as one can pass the eigenvalues when calling the function. If they do not pass anything, `generator_eigenval` will be `None` and qibo will calculate it by diagonalizing the matrix. Alternatively we could somehow hardcode the eigenvalues for Paulis in the code but I am not sure what is an elegant way to do this.

> By the way, point ii) applies not just to single-qubit Paulis, but also to any tensor product of Paulis; this made me think that perhaps we could implement a gate that is the exponential of a tensor product of Paulis times a variational parameter, I think it would be useful. 

I haven't done the math to ensure they are exactly the same but someone implemented such gates in #631 (already merged).

Reply:
As for (i), I agree with the solution that @stavros11  proposes for now. If the eigenvalues are explicitly given better, otherwise we can calculate them. Another course of action might be to have a dictionary that saves some notable generator-eigenvalue combinations. 

With your (ii) comment you mean a sequence of rotations in which the same parameter can be involved several times? e.g. `RY(theta) -> RZ -> RY -> RZ(theta)`  @DiegoGM91

Reply:
> I haven't done the math to ensure they are exactly the same but someone implemented such gates in https://github.com/qiboteam/qibo/pull/631 (already merged).

I see. This is certainly useful. I think this should be included in the documentation, I couldn't find any reference to them in the latest version of the documentation. Perhaps we could generalize these gates to general  imaginary exponentials of Pauli operators.

Also, it is easy to check that the eigenvalues of any tensor product of Pauli operators are always $\pm 1$, so the current parameter-shift rule can be applied to (imaginary) exponentials of them.


> With your (ii) comment you mean a sequence of rotations in which the same parameter can be involved several times? e.g. `RY(theta) -> RZ -> RY -> RZ(theta)` @DiegoGM91

Nope, I mean $e^{-i\theta P}$, where $P$ is a tensor product of Pauli gates, e.g. $P=\sigma^Z_1 \otimes \sigma^X_2 \otimes \sigma^Z_3$ where the subscripts indicate the qubit acted upon by each single-qubit Pauli. Note that in this example, this exponential would be a three-qubit gate.

Reply:
Do you think we can  open a new issue/PR combo for this further implementation and merge this PR before of that?
I purpose this in order to make it lighter. 
In this case, we can resolve this conversation and link it to the new issue/PR.